From 94c758eb859c1cc90880eb8ac8ce0d0325867347 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 5 Aug 2024 14:55:10 +0800
Subject: [PATCH 35/92] mtk: mt76: mt7996: add Adv-TTLM support for STA

1. add the handling for valid_link and TTLM changing in vif_cfg_changed
   callback.
2. send peer-mld request for default mapping and Adv-TTLM. Neg-TTLM will
   be supported in further commit.

mtk: mt76: mt7996: ignore vif.dormant_links in mt7996_change_vif_links

The dormant links are disabled for reasons like TTLM, and might become
active in the near future, so we do not remove their bss_conf here.

mtk: mt76: mt7996: pass vif->cfg.assoc to mt7996_mac_sta_add_links

Link management functions might leverage mt7996_change_sta_links to
update sta's links, so the 'assoc' argument of mt7996_mac_sta_add_links()
should be dynamic decieded by vif->cfg.assoc.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 mt76_connac_mcu.h |  1 +
 mt7996/main.c     |  4 ++++
 mt7996/mcu.h      | 22 ++++++++++++++++++++++
 mt7996/mt7996.h   |  2 ++
 4 files changed, 29 insertions(+)

diff --git a/mt76_connac_mcu.h b/mt76_connac_mcu.h
index 8721ecbd..a49dff79 100644
--- a/mt76_connac_mcu.h
+++ b/mt76_connac_mcu.h
@@ -1330,6 +1330,7 @@ enum {
 	MCU_UNI_CMD_ASSERT_DUMP = 0x6f,
 	MCU_UNI_CMD_PTA_3WIRE_CTRL = 0x78,
 	MCU_UNI_CMD_MLD = 0x82,
+	MCU_UNI_CMD_PEER_MLD = 0x83,
 	MCU_UNI_CMD_SDO = 0x88,
 };
 
diff --git a/mt7996/main.c b/mt7996/main.c
index 3d113939..5fa2a204 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -951,6 +951,10 @@ mt7996_vif_cfg_changed(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 		}
 	}
 
+	if ((changed & BSS_CHANGED_MLD_VALID_LINKS) &&
+	    (changed & (BSS_CHANGED_MLD_ADV_TTLM | BSS_CHANGED_MLD_NEG_TTLM)))
+		mt7996_mcu_peer_mld_ttlm_req(dev, vif, changed);
+
 	mutex_unlock(&dev->mt76.mutex);
 }
 
diff --git a/mt7996/mcu.h b/mt7996/mcu.h
index 6e353306..b0df9dea 100644
--- a/mt7996/mcu.h
+++ b/mt7996/mcu.h
@@ -1247,6 +1247,28 @@ enum {
 	UNI_EVENT_MLD_RECONF_AP_REM_TIMER = 0x04,
 };
 
+struct peer_mld_req_hdr {
+	u8 ver;
+	u8 peer_mld_addr[ETH_ALEN];
+	u8 mld_idx;
+	u8 rsv[4];
+	u8 buf[];
+} __packed;
+
+struct peer_mld_ttlm_req {
+	__le16 tag;
+	__le16 len;
+	u8 mld_addr[ETH_ALEN];
+	__le16 enabled_link_bitmap;
+	__le16 link_to_wcid[IEEE80211_MLD_MAX_NUM_LINKS + 1];
+	u8 dl_tid_map[IEEE80211_MLD_MAX_NUM_LINKS + 1];
+	u8 ul_tid_map[IEEE80211_MLD_MAX_NUM_LINKS + 1];
+} __packed;
+
+enum {
+	UNI_CMD_PEER_MLD_TTLM_REQ = 0x0,
+};
+
 struct tx_power_ctrl {
 	u8 _rsv[4];
 
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index 4fd3e3d2..a7c5ffd9 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -1172,6 +1172,8 @@ int mt7996_mcu_mld_link_oper(struct mt7996_phy *phy,
 			     struct mt7996_bss_conf *mconf, bool add);
 int mt7996_mcu_mld_set_attlm(struct mt7996_dev *dev, struct ieee80211_vif *vif,
 			     u16 disabled_links, u16 switch_time, u32 duration);
+int mt7996_mcu_peer_mld_ttlm_req(struct mt7996_dev *dev, struct ieee80211_vif *vif,
+				 u64 changed);
 
 static inline u8 mt7996_max_interface_num(struct mt7996_dev *dev)
 {
-- 
2.45.2

